targets/csharp/source/source/PlayFabGameServerInternal.cs

```diff
@@ -0,0 +1,310 @@
+                    }
+
+                    _signalHeartbeatEvent.Reset();
+                }
+
+                await SendHeartbeatAsync();
+            }
+
+            Logger.Log("Shutting down heartbeat thread");
+            _heartbeatDoneEvent.Set();
+        }
+
+        internal async virtual Task SendHeartbeatAsync()
+        {
+            bool isGameHealthy = false;
+            if (HealthCallback != null)
+            {
+                isGameHealthy = HealthCallback();
+            }
+
+            var payload = new HeartbeatRequest
+            {
+                CurrentGameState = this.State,
+                CurrentGameHealth = isGameHealthy ? "Healthy" : "Unhealthy",
+                CurrentPlayers = ConnectedPlayers.ToArray(),
+            };
+
+            try
+            {
+                HeartbeatResponse response = await _webClient.SendHeartbeatAsync(payload);
+                await UpdateStateFromHeartbeatAsync(response);
+
+                if (_shouldWriteDebugLogs)
+                {
+                    Logger.Log($"Heartbeat request: {{ state = {payload.CurrentGameState} }} response: {{ operation = {response.Operation} }}");
+                }
+            }
+            catch (Exception ex)
+            {
+                Logger.Log($"Cannot send heartbeat: {ex.Message}\r\n\r\n{ex}");
+                return;
+            }
+        }
+
+        private Task UpdateStateFromHeartbeatAsync(HeartbeatResponse response)
+        {
+            if (response.SessionConfig != null)
+            {
+                ConfigMap.AddIfNotNullOrEmpty("sessionCookie", response.SessionConfig?.SessionCookie);
+                ConfigMap.AddIfNotNullOrEmpty("sessionId", response.SessionConfig?.SessionId.ToString());
+
+                // Only setting InitialPlayers if something was sent from the Agent, this will prevent us effectively
+                // deleting our initial player list if the next heartbeat comes back empty.
+                if (response.SessionConfig?.InitialPlayers?.Count > 0)
+                {
+                    InitialPlayers = response.SessionConfig.InitialPlayers;
+                }
+            }
+
+            if (!string.IsNullOrWhiteSpace(response.NextScheduledMaintenanceUtc))
+            {
+                if (DateTime.TryParse(
+                    response.NextScheduledMaintenanceUtc,
+                    null,
+                    DateTimeStyles.RoundtripKind,
+                    out DateTime scheduledMaintDate))
+                {
+                    if (_cachedScheduleMaintDate == DateTime.MinValue || (scheduledMaintDate != _cachedScheduleMaintDate))
+                    {
+                        MaintenanceCallback?.Invoke(scheduledMaintDate);
+                        _cachedScheduleMaintDate = scheduledMaintDate;
+                    }
+                }
+            }
+
+            switch (response.Operation)
+            {
+                case GameOperation.Continue:
+                    {
```

Comment:
I am not a fan of wrapping cases in these extra brackets, as it's already implied.
It only serves to add lines, and tab in the code unnecessarily.